streaming/base/stream.py

```diff
@@ -0,0 +1,315 @@
+            the ``split`` subdirs of  ``remote`` and ``local``. Defaults to ``None``.
+        proportion (float, optional): How much to upsample or downsample this sub-dataset, as the
+            proportion of the total combined dataset that consists of this sub-dataset. If
+            using proportions, all sub-datasets provided together to the StreamingDataset init must
+            define their proportions. The total combined number of samples is either the
+            StreamingDataset argument "samples_per_epoch" if provided, or kept the same total size
+            as the underlying data if not. If provided, must be positive. Defaults to ``None``.
+        repeat (float, optional): How much to upsample or downsample this sub-dataset, as a
+            multipler on the number of samples. If provided, must be positive. Defaults to
+            ``None``.
+        samples (int, optional): How much to upsample or downsample this sub-dataset, as the exact
+            number of samples. If provided, must be positive. Defaults to ``None``.
+        download_retry (int, optional): Number of download re-attempts before giving up. Defaults
+            to ``None``.
+        download_timeout (float, optional): Number of seconds to wait for a shard to download
+            before raising an exception. Defaults to ``None``.
+        validate_hash (str, optional): Optional hash or checksum algorithm to use to validate
+            shards. Defaults to ``None``.
+        keep_zip (bool, optional): Whether to keep or delete the compressed form when decompressing
+            downloaded shards. If ``False``, keep iff remote is local or no remote. Defaults to
+            `None``.
+        keep_raw (bool, optional): Whether to keep or delete the decompressed form (or only form)
+            of shards after all their samples have been yielded this epoch. If ``False``, keep iff
+            remote is local or no remote and no compression. Defaults to ``None``.
+    """
+
+    def __init__(self,
+                 *,
+                 remote: Optional[str] = None,
+                 local: Optional[str] = None,
+                 split: Optional[str] = None,
+                 proportion: Optional[float] = None,
+                 repeat: Optional[float] = None,
+                 samples: Optional[int] = None,
+                 download_retry: Optional[int] = None,
+                 download_timeout: Optional[float] = None,
+                 validate_hash: Optional[str] = None,
+                 keep_zip: Optional[bool] = None,
+                 keep_raw: Optional[bool] = None) -> None:
+        self.remote = remote
+        self._local = local
+        self.local = local or mkdtemp()
+        self.split = split or ''
+
+        has_proportion = proportion is not None
+        has_repeat = repeat is not None
+        has_samples = samples is not None
+        if not (0 <= has_proportion + has_repeat + has_samples <= 1):
```

Comment:
Can we move all the validation checks to a separate method called `def _validate()` or something like that ? The same method can also be called in `StreamingDataset` `init()` method since we do some validation check there as well (For example, `if not (0 <= has_proportion + has_repeat + has_samples <= 1):` )!!

Reply:
eh, it's complex and wouldn't be able to be moved entirely to one side or the other...